False-positive for D208 "Docstring is over-indented" for code sample at the end of module docstring

I don't know if this appears anywhere else, but I observed it in a module docstring.

This:
```python
"""Module docstring.

.. code:: python
   a = f()
   assert a
"""
```
gives the error:
```
Found 1 error(s).
main.py:5:1: D208 Docstring is over-indented
```
whereas this:
```python
"""Module docstring.

.. code:: python
   a = f()
   assert a

More documentation.
"""
```
gives no error.

I can understand though if this is a wont-fix because I can just turn off the `D208` error code.

--- a/src/ast/whitespace.rs
+++ b/src/ast/whitespace.rs
@@ -1,3 +1,5 @@
+use std::str::Lines;
+
 use rustpython_ast::{Located, Location};
 
 use crate::ast::types::Range;
@@ -36,4 +38,39 @@ pub fn clean(indentation: &str) -> String {
         .chars()
         .map(|char| if char.is_whitespace() { char } else { ' ' })
         .collect()
+}
+
+/// Like `str#lines`, but includes a trailing newline as an empty line.
+pub struct LinesWithTrailingNewline<'a> {
+    trailing: Option<&'a str>,
+    underlying: Lines<'a>,
+}
+
+impl<'a> LinesWithTrailingNewline<'a> {
+    pub fn from(input: &'a str) -> LinesWithTrailingNewline<'a> {
+        LinesWithTrailingNewline {
+            underlying: input.lines(),
+            trailing: if input.ends_with('\n') {
+                Some("")
+            } else {
+                None
+            },
+        }
+    }
+}
+
+impl<'a> Iterator for LinesWithTrailingNewline<'a> {
+    type Item = &'a str;
+
+    #[inline]
+    fn next(&mut self) -> Option<&'a str> {
+        let mut next = self.underlying.next();
+        if next.is_none() {
+            if self.trailing.is_some() {
+                next = self.trailing;
+                self.trailing = None;
+            }
+        }
+        next
+    }
 }

--- a/src/pydocstyle/plugins.rs
+++ b/src/pydocstyle/plugins.rs
@@ -8,6 +8,7 @@ use rustpython_ast::{Constant, ExprKind, Location, StmtKind};
 
 use crate::ast::types::Range;
 use crate::ast::whitespace;
+use crate::ast::whitespace::LinesWithTrailingNewline;
 use crate::autofix::Fix;
 use crate::check_ast::Checker;
 use crate::checks::{Check, CheckCode, CheckKind};
@@ -126,7 +127,7 @@ pub fn one_liner(checker: &mut Checker, definition: &Definition) {
         {
             let mut line_count = 0;
             let mut non_empty_line_count = 0;
-            for line in string.lines() {
+            for line in LinesWithTrailingNewline::from(string) {
                 line_count += 1;
                 if !line.trim().is_empty() {
                     non_empty_line_count += 1;
@@ -136,7 +137,7 @@ pub fn one_liner(checker: &mut Checker, definition: &Definition) {
                 }
             }
 
-            if non_empty_line_count == 1 && line_count > 1 {
+            if non_empty_line_count == 1 && (line_count > 1) {
                 checker.add_check(Check::new(
                     CheckKind::FitsOnOneLine,
                     Range::from_located(docstring),
@@ -387,12 +388,14 @@ pub fn indent(checker: &mut Checker, definition: &Definition) {
             ..
         } = &docstring.node
         {
-            let lines: Vec<&str> = string.lines().collect();
+            // Split the docstring into lines.
+            let lines: Vec<&str> = LinesWithTrailingNewline::from(string).collect();
             if lines.len() <= 1 {
                 return;
             }
 
             let docstring_indent = whitespace::indentation(checker, docstring);
+
             let mut has_seen_tab = docstring_indent.contains('\t');
             let mut is_over_indented = true;
             let mut over_indented_lines = vec![];
@@ -418,7 +421,9 @@ pub fn indent(checker: &mut Checker, definition: &Definition) {
                 if checker.settings.enabled.contains(&CheckCode::D207) {
                     // We report under-indentation on every line. This isn't great, but enables
                     // autofix.
-                    if !is_blank && line_indent.len() < docstring_indent.len() {
+                    if (i == lines.len() - 1 || !is_blank)
+                        && line_indent.len() < docstring_indent.len()
+                    {
                         let mut check = Check::new(
                             CheckKind::NoUnderIndentation,
                             Range {
@@ -524,7 +529,7 @@ pub fn newline_after_last_paragraph(checker: &mut Checker, definition: &Definiti
         } = &docstring.node
         {
             let mut line_count = 0;
-            for line in string.lines() {
+            for line in LinesWithTrailingNewline::from(string) {
                 if !line.trim().is_empty() {
                     line_count += 1;
                 }
@@ -570,7 +575,7 @@ pub fn no_surrounding_whitespace(checker: &mut Checker, definition: &Definition)
             ..
         } = &docstring.node
         {
-            let mut lines = string.lines();
+            let mut lines = LinesWithTrailingNewline::from(string);
             if let Some(line) = lines.next() {
                 let trimmed = line.trim();
                 if trimmed.is_empty() {
@@ -627,7 +632,7 @@ pub fn multi_line_summary_start(checker: &mut Checker, definition: &Definition) 
             ..
         } = &docstring.node
         {
-            if string.lines().nth(1).is_some() {
+            if LinesWithTrailingNewline::from(string).nth(1).is_some() {
                 if let Some(first_line) = checker
                     .locator
                     .slice_source_code_range(&Range::from_located(docstring))
@@ -871,7 +876,7 @@ pub fn sections(checker: &mut Checker, definition: &Definition) {
             ..
         } = &docstring.node
         {
-            let lines: Vec<&str> = string.lines().collect();
+            let lines: Vec<&str> = LinesWithTrailingNewline::from(string).collect();
             if lines.len() < 2 {
                 return;
             }

